### PR TITLE
Add/hasoop request dispose is right and format data

### PR DESCRIPTION
**Pull request by smack3c0**

### CODE REVIEW

File: src/dispose.js [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+  const isOk = _.get(responseJson, 'api-versions[0]') === 'v1'
+  return {isRight: isOk, data: responseJson, headers: responseHeaders}
+}
+function getDriverDispose (responseJson, responseHeaders) {
+  const isOk = responseJson.version === '1' && _.get(responseJson, ['all-config-resources', 'jarConfig.label']) === 'Classpath configuration'
+  return {isRight: isOk, data: responseJson, headers: responseHeaders}
+}
+function getConnectorAllDispose (responseJson, responseHeaders) {
+  const connectorNames = _.map(responseJson.connectors, 'name')
+  const isOk = responseJson.connectors.length === 7 && _.includes(connectorNames, 'generic-jdbc-connector')
```
Comment (maicaiyao) [inline src/dispose.js]: why 7?
make it a constant with better name

File: src/dispose.js [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  // submission
+  startJob: null,
+  stopJob: null,
+  jobStatus: null,
+  getSubmissionAll: null,
+  getSubmissionByJobNam: null
+})
+
+function getVersionDispose (responseJson, responseHeaders) {
+  const isOk = _.get(responseJson, 'api-versions[0]') === 'v1'
```
Comment (maicaiyao) [inline src/dispose.js]: const isRight =
return { isRight, data: ... }

File: src/dispose.js [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+}
+function getLinkByLinkNameDispose (responseJson, responseHeaders, linkName) {
+  const linkConfig = splitLinkConfig(responseJson)
+  const linkConfigKeys = _.keys(linkConfig)
+  const isOk = linkConfig.name === linkName && _.includes(linkConfigKeys, 'id') && _.includes(linkConfigKeys, 'enabled') && _.includes(linkConfigKeys, 'connectorName')
+  const data = isOk ? linkConfig : responseJson
+  return {isRight: isOk, data, headers: responseHeaders}
+}
+function createLinkDispose (responseJson, responseHeaders, linkName) {
+  const isOk = _.isEqual(responseJson, {name: linkName, 'validation-result': [{}]})
```
Comment (maicaiyao) [inline src/dispose.js]: do we need deepEqual?

File: src/dispose.js [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+export async function hasoopRequestDispose (methodName, res, ...params) {
+  if (!Object.keys(hasoopMethodTypes).includes(methodName)) {
+    throw new Error(`hasoop method ${methodName} is not supported`)
+  }
+  const responseJson = await res.json()
+  const responseHeaders = getResponseHeaders(res)
+  if (!isRightFromHeaders(res)) {
+    return {isRight: false, data: responseJson, headers: responseHeaders}
+  }
+  switch (methodName) {
```
Comment (maicaiyao) [inline src/dispose.js]: better to have a function called getMethod and refactor the following code in it

const method = getMethod(methodName)
return method(responseJson, responseHeaders, ...params)

File: src/dispose.js [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  deleteJob: null,
+  deleteJobAll: null,
+  // submission
+  startJob: null,
+  stopJob: null,
+  jobStatus: null,
+  getSubmissionAll: null,
+  getSubmissionByJobNam: null
+})
+
```
Comment (maicaiyao) [inline src/dispose.js]: for the following method, I would create a function called getDisposeMethod

```
function disposeMethodGen (isRightGen, dataGen, headerGen) {
  (responseJson, responseHeaders, otherParams) => {
    const isRight = isRightGen(responseJson, otherParams)
    const data = dataGen(isRight, responseJson, otherParams)
    const headers = headerGen(responseHeaders, otherParams)
    return { isRight, data, headers }
  }
}
```

Comment (maicaiyao) [inline src/dispose.js]:
```
function versionIsRightGen (responseJson) {
  return _.get(responseJson, 'api-versions[0]') === 'v1'
}

function defaultDataGen (responseJson) {
  return responseJson
}

function getVersionDispose = disposeMethodGen(versionIsRightGen, defaultDataGen, xxx)
```

File: src/dispose.js [hunk 6]
```diff
@@ -2,6 +2,10 @@ import _ from 'lodash'
 import { simpleVersion, version, sqoopConnectorCount, hasoopMethodTypes } from './constant'
 import { splitLinkConfig, splitJobConfig, splitSubmissionConfig } from './utils'
 
+const equalEmptyObject = responseJson => _.isEqual(responseJson, {})
```
Comment (maicaiyao) [inline src/dispose.js]: isEmptyObject

Comment (maicaiyao) [inline src/dispose.js]: or just use _.isEmpty?

File: src/dispose.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+export const getResponseHeaders = res => ({
+  sqoopErrorCode: res.headers.get('sqoop-error-code'),
+  sqoopErrorMessage: res.headers.get('sqoop-error-message'),
+  sqoopInternalErrorCode: res.headers.get('sqoop-internal-error-code') || null,
+  sqoopInternalErrorMessage: res.headers.get('sqoop-internal-error-message') || null
+})
+
+export async function hasoopRequestDispose (methodName, res, ...params) {
+  if (!Object.keys(hasoopMethodTypes).includes(methodName)) {
+    throw new Error(`hasoop method ${methodName} is not supported`)
```
Comment (Jimexist) [inline src/dispose.js]: `is not supported` 和 `does not exist` 还是有区别的。你其实判断的是后者。

**Comment by maicaiyao** [comment]
ship it

**Pull request merged**